README
Have not had a chance to fill the README file with anything of value
yet.


ChangeLog
2005-12-14 Andrew De Ponte <cyphactor@socal.rr.com>

    * Created the plugin shell and the base CVS module so that
    development of the plugin could begin.


src/zdtm.c
/*
 * Copyright 2005-2006 Andrew De Ponte
 * 
 * This file is part of os_zdtm_plugin.
 * 
 * os_zdtm_plugin is free software; you can redistribute it and/or
 * modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * os_zdtm_plugin is distributed in the hopes that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with os_zdtm_plugin; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

/**
 * @file zdtm.c
 * @brief This is an implementation file for a Zaurus DTM support library.
 *
 * The zdtm.c file is an implementation for a Zaurus DTM based
 * synchronization, general support library. It provides all the
 * definitions for all the general functions used to perform the Zaurus
 * side of a synchronization.
 */

#include "zdtm.h"

/**
 * Listen for an incoming synchronization connection from a Zaurus.
 *
 * The zdtm_listen_for_zaurus function creates a socket and configures
 * it to listen for a synchronization connection from a Zaurus. Note:
 * This function does not handle accepting a connection from the Zaurus
 * it just creates a socket and puts it in the proper state so that a
 * Zaurus may make a connection to it, which will be backlogged and can
 * be handled at a later point in time by a function which handles
 * accepting a connection from a Zaurus.
 * @param cur_env Pointer to the current zdtm library environment.
 * @return An integer representing success (zero) or failure (non-zero).
 * @retval 0 Successfully listening for a Zaurus connection.
 * @retval -1 Failed to create a socket.
 * @retval -2 Failed to set the socket REUSEADDR option.
 * @retval -3 Failed to bind the address to the socket.
 * @retval -4 Failed to put socket into a listening state.
 */
int zdtm_listen_for_zaurus(zdtm_env *cur_env) {
    int retval;
    struct sockaddr_in servaddr;
    int reuse_set_flag = 1;

    cur_env->listenfd = socket(AF_INET, SOCK_STREAM, 0);
    if (cur_env->listenfd == -1) {
        perror("zdtm_listen_for_zaurus - socket");
        return -1;
    }

    // Here I set a socket option so that if the applications ends
    // prematurely the socket is not blocked by the TIME_WAIT state.
    retval = setsockopt(cur_env->listenfd, SOL_SOCKET, SO_REUSEADDR,
        (void *)&reuse_set_flag, sizeof(reuse_set_flag));
    if (retval == -1) {
        perror("zdtm_listen_for_zaurus - setsockopt");
        return -2;
    }

    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port = htons(DLISTPORT);

    retval = bind(cur_env->listenfd, (struct sockaddr *)&servaddr,
        sizeof(servaddr));
    if (retval == -1) {
        perror("zdtm_listen_for_zaurus - bind");
        return -3;
    }

    retval = listen(cur_env->listenfd, 1);
    if (retval == -1) {
        perror("zdtm_listen_for_zaurus - listen");
        return -4;
    }

    // Return success
    return 0;
}

/**
 * Handle a Zaurus connection.
 *
 * The zdtm_handle_zaurus_connection function handles a backlogged
 * Zaurus connection if one exists. If no backlogged Zaurus connection
 * exists then zdtm_handle_zaurus_connection will block waiting for a
 * Zaurus connection, at which point it will release after accepting the
 * connection from the Zaurus. Note: In order to handle a Zaurus
 * connection one must first call zdtm_listen_for_zaurus to be in the
 * correct state.
 * @param cur_env Pointer to the current zdtm library environment.
 * @return An integer representing success (zero) or failure (non-zero).
 * @retval 0 Successfully accepted a Zaurus connection.
 * @retval -1 Failed to accept a Zaurus connection.
 * @retval -2 Failed to convert client address to quad dot format.
 */
int zdtm_handle_zaurus_connection(zdtm_env *cur_env) {
    struct sockaddr_in clntaddr;
    char source_addr[16];
    socklen_t len;
    char *retval;

    memset(&clntaddr, 0, sizeof(clntaddr));
    len = sizeof(clntaddr);
    cur_env->connfd = accept(cur_env->listenfd,
        (struct sockaddr *)&clntaddr, &len);
    if (cur_env->connfd == -1) {
        perror("zdtm_handle_zaurus_connection - accept");
        return -1;
    }

    retval = (char *)inet_ntop(AF_INET, &clntaddr.sin_addr,
        source_addr, 16);
    if (retval == NULL) {
        perror("zdtm_handle_zaurus_connection - inet_ntop");
        return -2;
    }

    //printf("zdtm_handle_zaurus_connection: Received connection from \
    //    %s, port %d.\n", source_addr, ntohs(clntaddr.sin_port));

    // Return in success.
    return 0;
}

/**
 * Close the Zaurus connection.
 *
 * The zdtm_close_zaurus_conn function closes the Zaurus connection.
 * Note: In order for to close a zaurus connection a working zaurus
 * connection must already exist. Hence, the zdtm_listen_for_zaurus and
 * the zdtm_handle_zaurus_conn must have previously been called.
 * @param cur_env Pointer to the current zdtm library environment.
 * @return An integer representing success (zero) or failure (non-zero).
 * @retval 0 Successfully closed the Zaurus connection.
 * @retval -1 Failed to close the Zaurus connection.
 */
zdtm_close_zaurus_conn(zdtm_lib_env *cur_env) {
    int retval;

    retval = close(cur_env->connfd);
    if (retval) {
        perror("zdtm_close_zaurus_conn - close");
        return -1;
    }
    
    return 0;
}


src/zdtm.h
/*
 * Copyright 2005-2006 Andrew De Ponte
 * 
 * This file is part of os_zdtm_plugin.
 * 
 * os_zdtm_plugin is free software; you can redistribute it and/or
 * modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * os_zdtm_plugin is distributed in the hopes that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with os_zdtm_plugin; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

/**
 * @file zdtm.h
 * @brief This is a specifications file for a Zaurus DTM support library.
 *
 * The zdtm.h file is a specifications for a Zaurus DTM based
 * synchronization, general support library. It provides all prototypes
 * of the general functions used to perform the Zaurus side of a
 * synchronization.
 */

#ifndef ZDTM_H
#define ZDTM_H

// Standard Input/Output Includes
#include <stdio.h>

// Memory Includes
#include <string.h>

// Network Related Includes
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

// This is the port that the Zaurus listens on waiting for a connection
// to initiate a synchronization from the Desktop.
#define ZLISTPORT 4244
// This is the port that the Desktop synchronization server listens on
// waiting for a connection from the Zaurus to perform a synchronization.
#define DLISTPORT 4245

// This is the predefined constant size a Zaurus DTM Message header.
#define MSG_HDR_SIZE 13

/**
 * Zaurus DTM environment.
 *
 * The zdtm_lib_env is a type defined to represent the environment for a
 * given synchronization. It contains all the data that is required to
 * be shared among all the different zdtm library functions. Hence,
 * every synchronization must first have it's own non-shared (not shared
 * between synchronizations) environment before being able to use
 * the provided functions. Note: It does not have to be initialized in
 * any certain way, it can be filled with garbage as long as the
 * provided functions are called in the proper order.
 */
typedef struct zdtm_environment {
    int listenfd;
    int connfd;
} zdtm_lib_env;

/**
 * Zaurus DTM Message Body.
 *
 * The zdtm_message_body is a structure which represents the body of a
 * Zaurus DTM message.
 */
struct zdtm_message_body {
    unsigned char type[3];
    void *p_content;
};

/**
 * Zaurus DTM Message.
 *
 * The zdtm_message is a structure which represents a Zaurus DTM
 * Message.
 */
struct zdtm_message {
    char header[MSG_HDR_SIZE];
    struct zdtm_message_body body;
    unsigned short int body_size;
    unsigned short int check_sum;
    unsigned short int cont_size;
};

int zdtm_listen_for_zaurus(zdtm_lib_env *cur_env);
int zdtm_handle_zaurus_conn(zdtm_lib_env *cur_env);
int zdtm_close_zaurus_conn(zdtm_lib_env *cur_env);

#endif


src/format.c
/*
 * Copyright 2005-2006 Andrew De Ponte
 * 
 * This file is part of os_zdtm_plugin.
 * 
 * os_zdtm_plugin is free software; you can redistribute it and/or
 * modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * os_zdtm_plugin is distributed in the hopes that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with os_zdtm_plugin; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

/*
 * Here you can specify a plugin if you like.
 * You should use a format plugin, if the format conversion
 * is rather difficult or if other plugins might be able to reuse
 * your conversion
 * 
 */
#include <opensync/opensync.h>
#include "zdtm_sync.h"

static OSyncConvCmpResult compare_format1(OSyncChange *leftchange, OSyncChange *rightchange)
{
	/*
	 * This function can be used to compare two types of your formats.
	 * This is optional. For example, you can only provide a conversion
	 * from and to the xml format and let all the comparison be done there
	 * 
	 */

	/*
	 * Compare your objects here. You might need to cast the data of the change
	 * objects to you own defined types.
	 * 
	 * The possible result of the comparison are:
	 * 
	 * return CONV_DATA_SAME;
	 * The objects are exactly the same. (they might differ in some uid or
	 * timestamp though, but ALL the "real" information is the same)
	 * 
	 * return CONV_DATA_SIMILAR;
	 * The objects are not _exactly_ the same, but they look similar. This is used
	 * to detect conflicts. It is up to you to decide what "similar" means for your
	 * object
	 * 
	 * return CONV_DATA_MISMATCH;
	 * This means the objects are not the same and not similar.
	 * 
	 */
	return CONV_DATA_MISMATCH;
}

static osync_bool conv_format1_to_format2(void *conv_data, char *input, int inpsize, char **output, int *outpsize, osync_bool *free_input, OSyncError **error)
{
	/*
	 * This function can be used to convert your format to another format.
	 * Return TRUE if the conversion was successfull or return FALSE and set
	 * the error if something bad has happend.
	 * 
	 */
	
	/* The arguments mean:
	 * 
	 * - conv_data:
	 * Pointer to the data you returned in your init function (if any)
	 * 
	 * - input:
	 * The data you need to convert
	 * - inpsize
	 * The size of the input data
	 * 
	 * - output:
	 * After converting you need to set this
	 * to your result
	 * - outpsize:
	 * The size of the output
	 * 
	 * - free_input:
	 * You need to set this to TRUE if opensync
	 * can free the input after the conversion (so you dont
	 * use any reference from or to the input). A example where
	 * *free_input = FALSE; needs to be done would be a encapsulator
	 * that stores the input reference somewhere in its struct
	 * 
	 * - error:
	 * if something bad happens and you cannot convert, set the error!
	 * 
	 */
	
	return TRUE;
}

static osync_bool conv_format2_to_format1(void *conv_data, char *input, int inpsize, char **output, int *outpsize, osync_bool *free_input, OSyncError **error)
{
	/*
	 * This function can be used to convert another format to your format.
	 * Return TRUE if the conversion was successfull or return FALSE and set
	 * the error if something bad has happend.
	 * 
	 */
	
	return TRUE;
}

static void destroy_format1(char *input, size_t inpsize)
{
	/*
	 * Here you have to free the data allocated by your format
	 * 
	 */
}

static void duplicate_format1(OSyncChange *change)
{
	/*
	 * This function can be used to duplicate your format.
	 * Duplication does not mean to make 2 objects out of one,
	 * but to change to uid of the object in such a way that it
	 * differes from the original uid.
	 * 
	 * Most format will never need this.
	 * 
	 */
}

static char *print_format1(OSyncChange *change)
{
	/*
	 * If your format is not in a human printable format already
	 * you have to return a human readable string here describing the object
	 * as closely as possible. This information will be used by the user to decide
	 * which object to pick in a conflict.
	 * 
	 */
	 return NULL;
}

void get_info(OSyncEnv *env)
{
	/*
	 * Here you have to give opensync some information about your format
	 * This function will be called directly after the plugin has been loaded.
	 * 
	 */
	
	//Tell opensync for which object type you want to add a format.
	//If the object type does not exist yet, it will be created.
	osync_env_register_objtype(env, "<some object type>");
	
	//Tell opensync that we want to register a new format
	osync_env_register_objformat(env, "<some object type>", "<your format name>");
	//Now we can set the function on your format we have created above
	osync_env_format_set_compare_func(env, "<your format name>", compare_format1);
	osync_env_format_set_duplicate_func(env, "<your format name>", duplicate_format1);
	osync_env_format_set_destroy_func(env, "<your format name>", destroy_format1);
	osync_env_format_set_print_func(env, "<your format name>", print_format1);
	
	/*
	 * Now we can register the converters.
	 * You can specify 2 types of converters:
	 * 
	 * CONVERTER_CONV:
	 * This will specify a converter that directly converts between 2 formats
	 * 
	 * CONVERTER_ENCAP:
	 * This specifies a converter that "encapsulates" other data.
	 * Example:
	 * you can store a vcard inside a file. This actually means that you
	 * can store the vcard format inside the file format ("encapsulation")
	 * 
	 * CONVERTER_DESENCAP:
	 * like the example above, but in the other direction
	 * 
	 */
	osync_env_register_converter(env, CONVERTER_CONV, "<another format name>", "<your format name>", conv_format2_to_format1);
	osync_env_register_converter(env, CONVERTER_CONV, "<your format name>", "<another format name>", conv_format1_to_format2);
}


src/zdtm_sync.h
/*
 * Copyright 2005-2006 Andrew De Ponte
 * 
 * This file is part of os_zdtm_plugin.
 * 
 * os_zdtm_plugin is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * os_zdtm_plugin is distributed in the hopes that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with os_zdtm_plugin; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

#include "zdtm.h"

//Specify any structs etc here.

typedef struct plugin_environment {
    OSyncMember *member;
    //If you need a hashtable:
    OSyncHashTable *hashtable;
    //More stuff you need goes here

    // The Zaurus DTM environment
    struct zdtm_environment zdtm_env;
} zdtm_plugin_env;


src/zdtm_sync.c
/*
 * Copyright 2005-2006 Andrew De Ponte
 * 
 * This file is part of os_zdtm_plugin.
 * 
 * os_zdtm_plugin is free software; you can redistribute it and/or
 * modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 * 
 * os_zdtm_plugin is distributed in the hopes that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with os_zdtm_plugin; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

#include <opensync/opensync.h>
#include "zdtm_sync.h"
#include <stdlib.h>
#include <string.h>
#include <assert.h>

static void *zdtm_initialize(OSyncMember *member, OSyncError **error)
{
    char *configdata;
    int configsize;
    
    //You need to specify the <some name>_environment somewhere with
    //all the members you need
    zdtm_plugin_env *env = malloc(sizeof(zdtm_plugin_env));
    assert(env != NULL);
    memset(env, 0, sizeof(zdtm_plugin_env));
    
    //now you can get the config file for this plugin
    if (!osync_member_get_config(member, &configdata, &configsize, error)) {
        osync_error_update(error, "Unable to get config data: %s",
            osync_error_print(error));
        free(env);
        return NULL;
    }
    
    //Process the configdata here and set the options on your environment
    free(configdata);
    env->member = member;
    
    //If you need a hashtable you make it here
    env->hashtable = osync_hashtable_new();
    
    //Now your return your struct.
    return (void *)env;
}

static void zdtm_connect(OSyncContext *ctx)
{
    //Each time you get passed a context (which is used to track
    //calls to your plugin) you can get the data your returned in
    //initialize via this call:
    zdtm_plugin_env *env = (zdtm_plugin_env *)osync_context_get_plugin_data(ctx);

    /*
     * Now connect to your devices and report
     * 
     * an error via:
     * osync_context_report_error(ctx, ERROR_CODE, "Some message");
     * 
     * or success via:
     * osync_context_report_success(ctx);
     * 
     * You have to use one of these 2 somewhere to answer the context.
     * 
     */
    
    //If you are using a hashtable you have to load it here
    OSyncError *error = NULL;
    if (!osync_hashtable_load(env->hashtable, env->member, &error)) {
        osync_context_report_osyncerror(ctx, &error);
        return;
    }
    
    //you can also use the anchor system to detect a device reset
    //or some parameter change here. Check the docs to see how it works
    char *lanchor = NULL;
    //Now you get the last stored anchor from the device
    if (!osync_anchor_compare(env->member, "lanchor", lanchor))
        osync_member_set_slow_sync(env->member, "<object type to request a slow-sync>", TRUE);
}

static void zdtm_get_changeinfo(OSyncContext *ctx)
{
    zdtm_plugin_env *env = (zdtm_plugin_env *)osync_context_get_plugin_data(ctx);
    
    //If you use opensync hashtables you can detect if you need
    //to do a slow-sync and set this on the hastable directly
    //otherwise you have to make 2 function like "get_changes" and
    //"get_all" and decide which to use using
    //osync_member_get_slow_sync
    if (osync_member_get_slow_sync(env->member, "<object type>"))
        osync_hashtable_set_slow_sync(env->hashtable, "<object type>");

    /*
     * Now you can get the changes.
     * Loop over all changes you get and do the following:
     */
        char *data = NULL;
        //Now get the data of this change
        
        //Make the new change to report
        OSyncChange *change = osync_change_new();
        //Set the member
        osync_change_set_member(change, env->member);
        //Now set the uid of the object
        osync_change_set_uid(change, "<some uid>");
        //Set the object format
        osync_change_set_objformat_string(change, "<the format of the object>");
        //Set the hash of the object (optional, only required if you use hashtabled)
        osync_change_set_hash(change, "the calculated hash of the object");
        //Now you can set the data for the object
        //Set the last argument to FALSE if the real data
        //should be queried later in a "get_data" function
        
        osync_change_set_data(change, data, sizeof(data), TRUE);            

        //If you use hashtables use these functions:
        if (osync_hashtable_detect_change(env->hashtable, change)) {
            osync_context_report_change(ctx, change);
            osync_hashtable_update_hash(env->hashtable, change);
        }    
        //otherwise just report the change via
        //osync_context_report_change(ctx, change);

    //When you are done looping and if you are using hashtables    
    osync_hashtable_report_deleted(env->hashtable, ctx, "data");
    
    //Now we need to answer the call
    osync_context_report_success(ctx);
}

static osync_bool zdtm_commit_change(OSyncContext *ctx, OSyncChange *change)
{
    zdtm_plugin_env *env = (zdtm_plugin_env *)osync_context_get_plugin_data(ctx);
    
    /*
     * Here you have to add, modify or delete a object
     * 
     */
    switch (osync_change_get_changetype(change)) {
        case CHANGE_DELETED:
            //Delete the change
            //Dont forget to answer the call on error
            break;
        case CHANGE_ADDED:
            //Add the change
            //Dont forget to answer the call on error
            //If you are using hashtables you have to calculate the hash here:
            osync_change_set_hash(change, "new hash");
            break;
        case CHANGE_MODIFIED:
            //Modify the change
            //Dont forget to answer the call on error
            //If you are using hashtables you have to calculate the new hash here:
            osync_change_set_hash(change, "new hash");
            break;
        default:
            osync_debug("FILE-SYNC", 0, "Unknown change type");
    }
    //Answer the call
    osync_context_report_success(ctx);
    //if you use hashtable, update the hash now.
    osync_hashtable_update_hash(env->hashtable, change);
    return TRUE;
}

static void zdtm_sync_done(OSyncContext *ctx)
{
    zdtm_plugin_env *env = (zdtm_plugin_env *)osync_context_get_plugin_data(ctx);
    
    /*
     * This function will only be called if the sync was successfull
     */
    
    //If we have a hashtable we can now forget the already reported changes
    osync_hashtable_forget(env->hashtable);
    
    //If we use anchors we have to update it now.
    char *lanchor = NULL;
    //Now you get/calculate the current anchor of the device
    osync_anchor_update(env->member, "lanchor", lanchor);
    
    //Answer the call
    osync_context_report_success(ctx);
}

static void zdtm_disconnect(OSyncContext *ctx)
{
    zdtm_plugin_env *env = (zdtm_plugin_env *)osync_context_get_plugin_data(ctx);
    
    //Close all stuff you need to close
    
    //Close the hashtable
    osync_hashtable_close(env->hashtable);
    //Answer the call
    osync_context_report_success(ctx);
}

static void zdtm_finalize(void *data)
{
    zdtm_plugin_env *env = (zdtm_plugin_env *)data;
    //Free all stuff that you have allocated here.
    osync_hashtable_free(env->hashtable);

    free(env);
}

void get_info(OSyncEnv *env)
{
    //Now you can create a new plugin information and fill in the details
    //Note that you can create several plugins here
    OSyncPluginInfo *info = osync_plugin_new_info(env);
    
    //Tell opensync something about your plugin
    info->name = "zdtm-sync";
    info->longname = "Zaurus DTM Synchronization";
    info->description = "An OpenSync plugin which provides \
        synchronization with Zaurs Personal Mobal Tools (PMTs) which \
        are using a DTM based ROM (any of the recent Sharp ROMs).";
    //the version of the api we are using, (1 at the moment)
    info->version = 1;
    //If you plugin is threadsafe.
    info->is_threadsafe = TRUE;
    
    //Now set the function we made earlier
    info->functions.initialize = zdtm_initialize;
    info->functions.connect = zdtm_connect;
    info->functions.sync_done = zdtm_sync_done;
    info->functions.disconnect = zdtm_disconnect;
    info->functions.finalize = zdtm_finalize;
    info->functions.get_changeinfo = zdtm_get_changeinfo;
    
    //If you like, you can overwrite the default timeouts of your plugin
    //The default is set to 60 sec. Note that this MUST NOT be used to
    //wait for expected timeouts (Lets say while waiting for a webserver).
    //you should wait for the normal timeout and return a error.
    info->timeouts.connect_timeout = 5;
    //There are more timeouts for the other functions
    
    //Now you have to tell opensync all the object types that your are accepting.
    //There can be more than one type you accept (for example: contact, event, note etc)
    //osync_plugin_accept_objtype(info, "<object type name>");
    osync_plugin_accept_objtype(info, "todo");
    osync_plugin_accept_objtype(info, "contact");
    osync_plugin_accept_objtype(info, "event");
    //osync_plugin_accept_objtype(info, "note");
    
    //which format do you accept for this objtype
    //osync_plugin_accept_objformat(info, "<object type name>", "<format name>", "<name of the required extension if any>");
    osync_plugin_accept_objformat(info, "todo", "zdtm", NULL);
    osync_plugin_accept_objformat(info, "contact", "zdtm", NULL);
    osync_plugin_accept_objformat(info, "event", "zdtm", NULL);
    //osync_plugin_accept_objformat(info, "note", "zdtm", NULL)
    
    //set the commit function for this format. this function will be called for
    //each object to write once
    osync_plugin_set_commit_objformat(info, "<object type name>",
        "<format name>", zdtm_commit_change);
    osync_plugin_set_commit_objformat(info, "todo", "zdtm",
        zdtm_commit_change);
    osync_plugin_set_commit_objformat(info, "contact", "zdtm",
        zdtm_commit_change);
    osync_plugin_set_commit_objformat(info, "event", "zdtm",
        zdtm_commit_change);
    //osync_plugin_set_commit_objformat(info, "note", "zdtm",
    //    zdtm_commit_change);
    //the other possibility is to do a batch commit by setting
    //osync_plugin_set_batch_commit_objformat(info, "<object type name>", "<format name>", batch_commit);
    //this function will be called exactly once with all objects to write gathered in an array
}
